fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStore.java

```diff
@@ -1248,29 +1247,43 @@ public RecordCursor<Tuple> scanRecordKeys(@Nonnull final TupleRange range, @Null
                         .withSubspace(recordsSubspace)
                         .setContext(context)
                         .setContinuation(cont)
-                        .setLow(range.getLow(), range.getLowEndpoint())
-                        .setHigh(range.getHigh(), range.getHighEndpoint())
-                        .setScanProperties(scanProperties)
+                        .setRange(TupleRange.ALL)
+                        // inner cursor should not have return row limit
+                        .setScanProperties(scanProperties.with(ExecuteProperties::clearReturnedRowLimit))
                         .build();
                 // map the KV to the primary key
                 return cursor.map(kv -> {
                     final Tuple keyTuple = recordsSubspace.unpack(kv.getKey());
                     Tuple nextKey = keyTuple.popBack(); // Remove index item
-                    long nextIndex = keyTuple.getLong(keyTuple.size() - 1);
-                    // Some validation of the index to match known split enumerators
-                    if ((nextIndex != SplitHelper.RECORD_VERSION) && (nextIndex != SplitHelper.UNSPLIT_RECORD) && !(nextIndex >= SplitHelper.START_SPLIT_RECORD)) {
-                        throw new RecordCoreStorageException("Invalid record split number")
-                                .addLogInfo(LogMessageKeys.SPLIT_NEXT_INDEX,  nextIndex);
-                    }
+                    validateSuffix(keyTuple);
                     return nextKey;
                 });
             };
 
             return new DedupCursor<>(innerFunction, Tuple::fromBytes, Tuple::pack, continuation)
+                    // Apply row limit to the outer cursor
                     .limitRowsTo(scanProperties.getExecuteProperties().getReturnedRowLimit());
         }
     }
 
+    /**
+     * Internal utility to ensure the given key has a valid suffix.
+     * @param keyTuple the key tuple to validate
+     */
+    private void validateSuffix(final Tuple keyTuple) {
```

Comment:
Nit:
```suggestion
    private void validatePrimaryKeySuffix(final Tuple keyTuple) {
```

And perhaps it would be better of in `SplitHelper`

Reply:
Done